### ISSUE TITLE
Updates to Listener Port field

### ISSUE THREAD

**Issue by youngnick** [kubernetes-sigs/gateway-api#841]
As part of the v1alpha2 rewiew, @thockin noted that, for the Port field on Gateway Listeners,  it might be a good idea to consider:
- making the field optional
- extending the field into a struct to enable capturing use cases like "all ports" and "port range" in the future.

So, this issue is to discuss some questions around this:
- Should we make the Port field optional?
- How can we add functionality to handle cases like "all ports", "port range" (and possibly "no port") later?

**Why this is needed**:

We need to decide this as part of v1lapha2 so that we can get any breaking changes required into this version.

**Comment by youngnick** [comment]
My opinion is that we should make the port field optional, and clearly specify the expected behavior when it's not specified.

When we specify that behavior, the specification will need to be by protocol and/or Route type attached to the Listener. This means that we can also handle future use cases by definite the "absent" behavior in those cases. I think the definition should look something like this:

When a port is not specified, implementations are expected to react in the way that makes the most sense for the associated protocol or Route types. For example, with the four basic Gateway API types, when no port is specified:
- For HTTPRoute, if no TLS is specified, the port must be set to 80.
- For HTTPRoute if TLS is specified, the port must be set to 443.
- For TLSRoute, the port must be set to 443 (since most implementations will probably be using HTTP).
- For TCPRoute and UDPRoute, the port not being specified is an error, and the listener must be rejected. (I considered saying "no port == all ports", but this seems like a very insecure option).

I've started out very strict here, because I think it's easier to loosen this if we need to later, but in this early stage, as with many things, we need to ensure we're buliding the right expectation for users about what default behaviors are. I think that even this is pretty complicated, and adding the ability for implementations to pick a different default is making it more so.

**Comment by jpeach** [comment]
IIRC that there is a separate issue that says TLSRoute should should be valid with HTTPS.

So, for defaulting, if protocol is "HTTP" then the port can be defaulted to 80, and if the protocol is "HTTPS" it can be defaulted to 433. This seems like a pretty marginal benefit that is offset by the complexity of only sometime getting to omit the port. I don't think that it's worth it.

**Comment by howardjohn** [comment]
I am a bit confused why we are looking at the Route types, and not the listener.protocol? The port must be determined in absence of routes entirely IMO - you may have 0 routes (or switch from 0 -> 1 -> 0 -> 1, possibly of different types, etc). It seems like it should be:

* protocol=HTTP -> 80
* protocol=HTTPS -> 443
* Everything else - required

**Comment by ccfishk** [comment]
From what we have been discussed,  port is needed and can be set with default value, otherwise please specify,
Default (w/o specification)
-  HTTP->80
- HTTPS->443
Based on this, "all ports" can be applied to HTTP/HTTPS, or we can add a field "- ApplyToAll" which applying this ports to all services. 

If we are allowing same ports, we are asking unique listener names implicitly if I understand correctly.

**Comment by robscott** [comment]
> we should make the port field optional

Agreed.

> The port must be determined in absence of routes entirely IMO

Agreed.

Also agree that we should default to the following port values based on protocol:

> protocol=HTTP -> 80
> protocol=HTTPS -> 443

I think we should request that implementations automatically assign ports for other protocols when they are not specified. Although a small difference, that seems more consistent with HTTP(S) and simpler than requiring users to specify a port, especially if they're just looking for any port that isn't currently in use by the Gateway.

**Comment by howardjohn** [comment]
> I think we should request that implementations automatically assign ports for other protocols when they are not specified. Although a small difference, that seems more consistent with HTTP(S) and simpler than requiring users to specify a port, especially if they're just looking for any port that isn't currently in use by the Gateway.

I am pretty skeptical of this. We (or at least I) haven't heard any clear use case for this. It will required an additional API change to status, since now we need to report to the user which port we assigned it to. It will additionally require implementations to keep state about what port they assigned, and for in cluster implementations will require that the controller owns the Service (so they can update the port, etc).

**Comment by bowei** [comment]
Optional port: +1

Defaults: this is a SHOULD I think  -- 

* it is recommended that implementations assign standard ports (HTTP=>80, HTTPS=>443), 
  but we leave it open to not have this as a strict requirement.
* It is not a requirement that the implementation assign a free port or perform port management.
  Implementations are free can do so if they want.

**Comment by robscott** [comment]
> I am pretty skeptical of this. We (or at least I) haven't heard any clear use case for this.

I think there's a few contributing factors here:

1. We'll need to make port optional at some point to enable more advanced concepts like all ports, or port ranges.
2. Since the majority of our implementations our HTTP(S) focused, port is easy to derive from protocol
3. This API is already complicated, taking a required field and making it optional means there's one less thing users have to think about

So although I don't think we've had any users directly ask for auto-assigning ports, I think it's a fairly natural outcome.

> It will required an additional API change to status, since now we need to report to the user which port we assigned it to.

This is something we had in v1alpha1 but removed in v1alpha2 when we added name to listeners. So this will actually result in a slightly smaller diff with v1alpha1.

> For in cluster implementations will require that the controller owns the Service

Another way of looking at this could be that the linked Service included a list of allocatable ports, and if there were no allocatable ports remaining, that could be an error state.

> It is not a requirement that the implementation assign a free port or perform port management.

I can understand that not every implementation may be capable of port assignment, but I think we should encourage that pattern, and if a controller chooses not to implement it, they MUST add a warning condition to status if a listener does not specify a port and it will not be auto-assigned.

**Comment by howardjohn** [comment]
> So although I don't think we've had any users directly ask for auto-assigning ports, I think it's a fairly natural outcome.

I think I just fundamentally think we should make the functionality of the API match use cases, not add highly complex features just because they fit the API shape. Its OK to have the port optional for HTTP(S) but required for anything else. Auto-allocation is not a requirement for optional ports.

**Comment by jpeach** [comment]
Given the limited use cases for defaulting the port, my preference is to keep it required.

**Comment by youngnick** [comment]
I agree with making the port optional now, because I think that there are plenty of use cases where it's not necessary.  I think the future proofing reasons are the reason to do this action *now* rather than leaving it as required.

HTTP and HTTPS are the major use cases for now, by expected percentage of usage, so the UX benefit of not needing to specify default ports is nice.

I think that for the other listener types, we should document the expected behavior, something like:
```
An implementation MAY allocate unspecified ports automatically, or it MAY reject unspecified ports as invalid.

In any case, if a port is chosen through whatever mechanism, it MUST be listed in the status.whatever field, and if a port is rejected, then a condition of whatever must be raised with a reason of whatever.
```

**Comment by jpeach** [comment]
> I agree with making the port optional now, because I think that there are plenty of use cases where it's not necessary. I think the future proofing reasons are the reason to do this action _now_ rather than leaving it as required.
> 
> HTTP and HTTPS are the major use cases for now, by expected percentage of usage, so the UX benefit of not needing to specify default ports is nice.
> 
> I think that for the other listener types, we should document the expected behavior, something like:
> 
> ```
> An implementation MAY allocate unspecified ports automatically, or it MAY reject unspecified ports as invalid.
> 
> In any case, if a port is chosen through whatever mechanism, it MUST be listed in the status.whatever field, and if a port is rejected, then a condition of whatever must be raised with a reason of whatever.
> ```

Right not, users of the API can be completely clear on what ports do, and that's __portable__ across implementations. This proposal makes omitting the port legal, but also __unportable__. So, I still struggle to see that this is worth the cost. The benefit seems tiny to me.

**Comment by youngnick** [comment]
I think the main reason to make the port optional for me is to give us room to expand easily later. If it's required now, and we want to add in allPorts, noPorts, portRanges or similar later, then it's much harder to migrate than if the port is optional. We are pretty confident that some or all of those three cases *will need to be addressed at some point*, definitely. So I'd argue that keeping the port required for now is almost certainly building in a decision that will be difficult to fix later. 

tl;dr it's not opnmly about making it optional now just for brevity reasons, it's also about making it easier to add extra stuff later.

If we do keep this required, how are we going to address those use cases in the future?

**Comment by jpeach** [comment]
> If we do keep this required, how are we going to address those use cases in the future?

Make the port optional at that time (that's compatible change), and add a field for the new use cases?

**Comment by youngnick** [comment]
Fair enough. I see where you're coming from better now. 

So, we've considered making the port optional and adding extra things and said "not right now". So should we leave this open for when we do want to do that or close it?

**Comment by jpeach** [comment]
> Fair enough. I see where you're coming from better now.
> 
> So, we've considered making the port optional and adding extra things and said "not right now". So should we leave this open for when we do want to do that or close it?

I don't have a strong preference, but would lean toward opening a new issue later.

**Comment by youngnick** [comment]
So, we already have #818 to cover the "all ports" and similar use cases, so I will close this one out and suggest that we do further work there.

To summarise, we discussed and considered, and have decided *not* to make the Port field optional at this time. We will consider if we make the field optional when we decide what we will do about multiple-ports/no port/all ports functionality in #818.

**Comment by thockin** [comment]
Jumping in this late...

It's POSSIBLE to move required->optional but it is fraught.   Consider version skew.  Suppose  we release an API which makes that change (and we can do it in a rollback-safe way).  How does that impact controller implementations?  How tolerant do they have to be of fields which are nominally required actually being unspecified?

I'd love to see GW fit into the place where Service NodePorts and LBs live, and that requires allocating ports (NodePorts) and handling whole-IPs (optional protocol and port) or ranges.

Before Gateway API goes GA (and maybe beta?) I'd like to see us have confidence in how that would work (or assert that it won't work and take it off my brain-backlog).
